File: app/src/main/java/com/example/lettersidebar/d/LetterSideBar.java
package com.example.lettersidebar.d;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;

import com.example.lettersidebar.util;


public class LetterSideBar extends View {

    private Paint mPaint, mTouchPaint;

    private int mTextColor, mTouchColor;

    private int mTextSize;
    private char a = 'A';
    //当前触摸的位置的字母
    private char mCurrentTouchLetter;


    public LetterSideBar(Context context) {
        this(context, null);
    }

    public LetterSideBar(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LetterSideBar(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mTextColor = Color.BLUE;
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        //自定义属性
        mPaint.setTextSize(util.sp2px(context, 20));
        mPaint.setColor(mTextColor);

        mTouchPaint = new Paint();
        mTouchPaint.setColor(Color.RED);
        mTouchPaint.setAntiAlias(true);
        mTouchPaint.setTextSize(util.sp2px(context, 20));

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        //计算宽度  字母宽度(画笔）+padding

        int textWidth = (int) mPaint.measureText("A");
        int width = textWidth + getPaddingLeft() + getPaddingRight();
        //高度直接获取
        int height = MeasureSpec.getSize(heightMeasureSpec);

        setMeasuredDimension(width, height);

    }

    @Override
    protected void onDraw(Canvas canvas) {

        int itemHeight = (getHeight() - getPaddingBottom() - getPaddingTop()) / 26;
        char c;
        int x;
        for (int i = 0; i < 26; i++) {
            int letterCenterY = itemHeight / 2 + i * itemHeight + getPaddingTop();
            Paint.FontMetrics metrics = mPaint.getFontMetrics();

            int dy = (int) ((metrics.bottom - metrics.top) / 2 - metrics.bottom);
            int baseLine = letterCenterY + dy;

            c = (char) (a + i);
            x = (int) ((getWidth() - mPaint.measureText(String.valueOf(c))) / 2);

            //优化，最好用两个画笔

            if (c == mCurrentTouchLetter) {
                canvas.drawText(String.valueOf(c), x, baseLine, mPaint);
            } else {
                canvas.drawText(String.valueOf(c), x, baseLine, mTouchPaint);
            }

        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int itemHeight = (getHeight() - getPaddingBottom() - getPaddingTop()) / 26;
        switch (event.getAction()) {
            case MotionEvent.ACTION_MOVE:
            case MotionEvent.ACTION_DOWN:

                float currentMoveY = event.getY();
                float currentPosition = (currentMoveY - getPaddingTop()) / itemHeight;

                if (currentPosition < 0) {
                    currentPosition = 0;
                }
                if (currentPosition > 25) {
                    currentPosition = 25;
                }
                char touchLetter = (char) (a + currentPosition);

                //优化，不重复绘制
                if (mCurrentTouchLetter == touchLetter) {
                    return true;
                } else mCurrentTouchLetter = touchLetter;

                if (letterTouchListener != null) {
                    letterTouchListener.touch(mCurrentTouchLetter, true);
                }
                invalidate();
                break;
            case MotionEvent.ACTION_UP:
                if (letterTouchListener != null) {
                    letterTouchListener.touch(mCurrentTouchLetter, false);
                }
                break;
        }
        return true;
    }

    private LetterTouchListener letterTouchListener;

    public void setLetterTouchListener(LetterTouchListener letterTouchListener) {
        this.letterTouchListener = letterTouchListener;
    }

    public interface LetterTouchListener {
        void touch(char letter, boolean ifTouch);
    }
}

File: app/src/main/java/com/example/lettersidebar/LetterSideBar.kt
package com.example.lettersidebar

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View


class LetterSideBar  constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) :
    View(context, attrs, defStyleAttr) {
    private val mPaint: Paint
    private val mTouchPaint: Paint
    private val mTextColor: Int
    private val mTouchColor = 0
    private val mTextSize = 0
    private val a = 'A'

    //当前触摸的位置的字母
    private var mCurrentTouchLetter = 0.toChar()
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        //计算宽度  字母宽度(画笔）+padding
        val textWidth = mPaint.measureText("A").toInt()
        val width = textWidth + paddingLeft + paddingRight
        //高度直接获取
        val height = MeasureSpec.getSize(heightMeasureSpec)
        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        val itemHeight = (height - paddingBottom - paddingTop) / 26
        var c: Char
        var x: Int
        for (i in 0..25) {
            val letterCenterY = itemHeight / 2 + i * itemHeight + paddingTop
            val metrics = mPaint.fontMetrics
            val dy = ((metrics.bottom - metrics.top) / 2 - metrics.bottom).toInt()
            val baseLine = letterCenterY + dy
            c = (a + i).toChar()
            x = ((width - mPaint.measureText(c.toString())) / 2).toInt()

            //优化，最好用两个画笔
            if (c == mCurrentTouchLetter) {
                canvas.drawText(c.toString(), x.toFloat(), baseLine.toFloat(), mPaint)
            } else {
                canvas.drawText(c.toString(), x.toFloat(), baseLine.toFloat(), mTouchPaint)
            }
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val itemHeight = (height - paddingBottom - paddingTop) / 26
        when (event.action) {
            MotionEvent.ACTION_MOVE, MotionEvent.ACTION_DOWN -> {
                val currentMoveY = event.y
                var currentPosition = (currentMoveY - paddingTop) / itemHeight
                if (currentPosition < 0) {
                    currentPosition = 0f
                }
                if (currentPosition > 25) {
                    currentPosition = 25f
                }
                val touchLetter: Char = a + currentPosition.toInt()

                //优化，不重复绘制
                mCurrentTouchLetter = if (mCurrentTouchLetter == touchLetter) {
                    return true
                } else touchLetter
                if (letterTouchListener != null) {
                    letterTouchListener!!.touch(mCurrentTouchLetter, true)
                }
                invalidate()
            }
            MotionEvent.ACTION_UP -> if (letterTouchListener != null) {
                letterTouchListener!!.touch(mCurrentTouchLetter, false)
            }
        }
        return true
    }

    private var letterTouchListener: LetterTouchListener? = null

    init {
        mTextColor = Color.BLUE
        mPaint = Paint()
        mPaint.isAntiAlias = true
        //自定义属性
        mPaint.textSize = util.sp2px(context, 20).toFloat()
        mPaint.color = mTextColor
        mTouchPaint = Paint()
        mTouchPaint.color = Color.RED
        mTouchPaint.isAntiAlias = true
        mTouchPaint.textSize = util.sp2px(context, 20).toFloat()
    }

    fun setLetterTouchListener(letterTouchListener: LetterTouchListener?) {
        this.letterTouchListener = letterTouchListener
    }

    interface LetterTouchListener {
        fun touch(letter: Char, ifTouch: Boolean)
    }
}